proposals/module-initializers.md

```diff
@@ -0,0 +1,181 @@
+# Module Initializers
+
+* [x] Proposed
+* [ ] Prototype: [In Progress](https://github.com/jnm2/roslyn/tree/module_initializer)
+* [ ] Implementation: [In Progress](https://github.com/dotnet/roslyn/tree/features/module-initializers)
+* [ ] Specification: [Not Started]()
+
+## Summary
+[summary]: #summary
+
+Although the .NET platform has a feature that directly supports writing initialization code for the assembly (technically, the module), it is not exposed in C#.  This is a rather niche scenario, but once you run into it the solutions appear to be pretty painful.  There are reports of [a number of customers](https://www.google.com/search?q=.net+module+constructor+c%23&oq=.net+module+constructor) (inside and outside Microsoft) struggling with the problem, and there are no doubt more undocumented cases.
+
+## Motivation
+[motivation]: #motivation
+
+- Enable libraries to do eager, one-time initialization when loaded, with minimal overhead and without the user needing to explicitly call anything
+- One particular pain point is that the runtime must do additional checks on usage of a type with a static constructor, in order to decide whether the static constructor needs to be run or not. This adds measurable overhead.
+- Enable source generators to run some global initialization logic without the user needing to explicitly call anything
+  - (in which scenarios is a static constructor insufficient/undesirable for this?)
+
+## Detailed design
+[design]: #detailed-design
+
+Add a tiny feature to support this without any explicit syntax, by having the C# compiler recognize a module attribute with a well-known name, like the following:
+
+``` c#
+namespace System.Runtime.CompilerServices
+{
+    // Note: an Obsolete attribute is not needed here,
+    // because only C# 9 compilers will have access to attributes added in .NET 5.
+    // LangVersion checks will still be necessary.
+    [AttributeUsage(AttributeTargets.Module, AllowMultiple = false)]
+    public class ModuleInitializerAttribute : Attribute
+    {
+        public ModuleInitializerAttribute(Type type) { }
+    }
+}
+```
+
+You would use it like this
+
+``` c#
+using System.Runtime.CompilerServices;
+
+[module: ModuleInitializer(typeof(MyModuleInitializer))]
+
+internal static class MyModuleInitializer
+{
+    static MyModuleInitializer()
+    {
+        // put your module initializer here
+    }
+}
+```
+
+and the C# compiler would then emit a module constructor that causes the static constructor of the identified type to be triggered:
```

Comment:
This seems counter to one of the primary motivations of the feature:
> - One particular pain point is that the runtime must do additional checks on usage of a type with a static constructor, in order to decide whether the static constructor needs to be run or not. **This adds measurable overhead.**

We should expand why this is an acceptable solution.

Reply:
With this approach you would generally declare a type just for the purpose of holding your module initializer, and avoid referencing it outside of the `[ModuleInitializerAttribute]`. Added language to clarify this.